File: native/pact_consumer_nif/src/builders/interaction_builder.rs
use crate::builders::request_builder::NifRequestBuilder;
use crate::builders::response_builder::NifResponseBuilder;
use crate::impl_builder_nif;
use crate::models::interaction::InteractionResource;
use crate::models::interaction::NifInteraction;
use pact_consumer::builders::InteractionBuilder;
use rustler::{NifResult, NifStruct, Resource, ResourceArc};
use std::sync::Mutex;

#[derive(NifStruct)]
#[module = "InteractionBuilder"]
pub struct NifInteractionBuilder {
    inner: ResourceArc<InteractionBuilderResource>,
}

pub struct InteractionBuilderResource(Mutex<InteractionBuilder>);

impl NifInteractionBuilder {
    fn invoke<F, T>(&self, fun: F) -> NifResult<T>
    where
        F: FnOnce(&mut InteractionBuilder) -> NifResult<T>,
    {
        let mut inner = self
            .inner
            .0
            .lock()
            .map_err(|_e| rustler::Error::RaiseAtom("invalid_interaction_builder_reference"))?;

        fun(&mut inner)
    }
}

impl Resource for InteractionBuilderResource {}

#[rustler::nif(name = "interaction_builder_new")]
pub fn new(description: String, interaction_type: String) -> NifInteractionBuilder {
    NifInteractionBuilder {
        inner: ResourceArc::new(InteractionBuilderResource(Mutex::new(
            InteractionBuilder::new(description, interaction_type),
        ))),
    }
}

#[rustler::nif(name = "interaction_builder_given_with_params")]
pub fn given_with_params(
    builder: NifInteractionBuilder,
    given: String,
    params: String,
) -> NifResult<NifInteractionBuilder> {
    builder.invoke(|b| {
        let params: serde_json::Value = serde_json::from_str(&params)
            .map_err(|_e| rustler::Error::RaiseAtom("invalid_params"))?;

        b.given_with_params(given, &params);
        Ok(())
    })?;

    Ok(builder)
}

#[rustler::nif(name = "interaction_builder_request")]
pub fn request(
    builder: NifInteractionBuilder,
    request_builder: NifRequestBuilder,
) -> NifResult<NifInteractionBuilder> {
    builder.invoke(|b| {
        b.request = request_builder.0.lock().unwrap().clone();
        Ok(())
    })?;
    Ok(builder)
}

#[rustler::nif(name = "interaction_builder_response")]
pub fn response(
    builder: NifInteractionBuilder,
    response_builder: NifResponseBuilder,
) -> NifResult<NifInteractionBuilder> {
    builder.invoke(|b| {
        b.response = response_builder.0.lock().unwrap().clone();
        Ok(())
    })?;
    Ok(builder)
}

#[rustler::nif(name = "interaction_builder_build")]
pub fn build(builder: NifInteractionBuilder) -> NifResult<NifInteraction> {
    let interaction = builder.invoke(|b| Ok(b.build()))?;

    Ok(NifInteraction {
        inner: ResourceArc::new(InteractionResource(Box::new(interaction))),
    })
}

#[rustler::nif(name = "interaction_builder_build_v4")]
pub fn build_v4(builder: NifInteractionBuilder) -> NifResult<NifInteraction> {
    let interaction = builder.invoke(|b| Ok(b.build_v4()))?;

    Ok(NifInteraction {
        inner: ResourceArc::new(InteractionResource(Box::new(interaction))),
    })
}

impl_builder_nif!("interaction_builder_with_key", NifInteractionBuilder, with_key(key: String));
impl_builder_nif!("interaction_builder_pending", NifInteractionBuilder, pending(value: bool));
impl_builder_nif!("interaction_builder_given", NifInteractionBuilder, given(value: String));
impl_builder_nif!("interaction_builder_comment", NifInteractionBuilder, comment(value: String));
impl_builder_nif!("interaction_builder_test_name", NifInteractionBuilder, test_name(value: String));
impl_builder_nif!("interaction_builder_transport", NifInteractionBuilder, transport(name: String));

File: lib/native/builders/interaction_builder.ex
defmodule Pact.Native.Builders.InteractionBuilder do
  @moduledoc false

  alias Pact.Native.PactConsumer.Interaction

  defmacro __using__(_opts) do
    quote do
      defmodule InteractionBuilder do
        @moduledoc """
        Interaction Builder
        """
        @enforce_keys [:inner]
        defstruct [:inner]

        @type t :: %__MODULE__{
                inner: reference()
              }
      end

      @spec interaction_builder_new(description :: String.t(), interaction_type :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_new(_description, _interaction_type),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_with_key(buider :: InteractionBuilder.t(), key :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_with_key(_builder, _key), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_pending(builder :: InteractionBuilder.t(), pending :: bool()) ::
              InteractionBuilder.t()
      def interaction_builder_pending(_builder, _pending), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_given(builder :: InteractionBuilder.t(), given :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_given(_builder, _given), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_given_with_params(
              builder :: InteractionBuilder.t(),
              given :: String.t(),
              params :: String.t()
            ) :: InteractionBuilder.t()
      def interaction_builder_given_with_params(_builder, _given, _params),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_comment(builder :: InteractionBuilder.t(), comment :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_comment(_builder, _comment), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_test_name(builder :: InteractionBuilder.t(), name :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_test_name(_builder, _name), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_transport(builder :: InteractionBuilder.t(), name :: String.t()) ::
              InteractionBuilder.t()
      def interaction_builder_transport(_builder, _name), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_request(
              builder :: InteractionBuilder.t(),
              request_builder :: InteractionBuilder.t()
            ) ::
              InteractionBuilder.t()
      def interaction_builder_request(_builder, _request_builder),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_response(
              builder :: InteractionBuilder.t(),
              response_builder :: InteractionBuilder.t()
            ) ::
              InteractionBuilder.t()
      def interaction_builder_response(_builder, _response_builder),
        do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_build(builder :: InteractionBuilder.t()) :: Interaction.t()
      def interaction_builder_build(_builder), do: :erlang.nif_error(:nif_not_loaded)

      @spec interaction_builder_build_v4(builder :: InteractionBuilder.t()) :: Interaction.t()
      def interaction_builder_build_v4(_builder), do: :erlang.nif_error(:nif_not_loaded)
    end
  end
end
